prov/mrail/src/mrail_cq.c

```diff
@@ -30,8 +30,96 @@
  * SOFTWARE.
  */
 
+#include <ofi_iov.h>
+
 #include "mrail.h"
 
+#define MRAIL_OP_CTX_RAIL_EP(op_context)	\
+	(((struct fi_recv_context *)op_context)->ep)
+
+#define MRAIL_OP_CTX_EP(op_context)	\
+	((struct mrail_ep *)MRAIL_OP_CTX_RAIL_EP(op_context)->fid.context)
+
+int mrail_cq_process_buf_recv(mrail_cq_entry_t *comp, struct mrail_recv *recv)
+{
+	struct fi_msg msg = {
+		.context = comp->op_context,
+	};
+	struct mrail_ep *mrail_ep = MRAIL_OP_CTX_EP(comp->op_context);
+	struct mrail_pkt *mrail_pkt = (struct mrail_pkt *)comp->buf;
+	size_t size, len = comp->len - sizeof(*mrail_pkt);
+	int ret, retv = 0;
+
+	size = ofi_copy_to_iov(recv->iov, recv->count, 0, mrail_pkt->data, len);
+
+	if (size < len) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ, "Message truncated recv buf "
+			"size: %zu message length: %zu\n", size, len);
+		ret = ofi_cq_write_error_trunc(
+			mrail_ep->util_ep.rx_cq, recv->context,
+			recv->comp_flags | (comp->flags & FI_REMOTE_CQ_DATA),
+			0, NULL, comp->data, mrail_pkt->hdr.tag, comp->len - size);
+		if (ret) {
+			FI_WARN(&mrail_prov, FI_LOG_CQ,
+				"Unable to write truncation error to util cq\n");
+			retv = ret;
+		}
+		goto out;
+	}
+	FI_DBG(&mrail_prov, FI_LOG_CQ, "writing recv completion: length: %zu "
+	       "tag: 0x%" PRIx64 "\n", len, comp->tag);
+	ret = ofi_cq_write(mrail_ep->util_ep.rx_cq, recv->context,
+			   recv->comp_flags |
+			   (comp->flags & FI_REMOTE_CQ_DATA), len, NULL,
+			   comp->data, comp->tag);
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ, "Unable to write to util cq\n");
+		retv = ret;
+	}
+out:
+	ret = fi_recvmsg(MRAIL_OP_CTX_RAIL_EP(comp->op_context),
+			 &msg, FI_DISCARD);
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Unable to discard buffered recv\n");
+		retv = ret;
+	}
+	mrail_push_recv(mrail_ep, recv);
+	return retv;
+}
+
+static int mrail_cq_process_comp_buf_recv(struct util_cq *cq,
+					  mrail_cq_entry_t *comp)
+{
+	struct mrail_ep *mrail_ep = MRAIL_OP_CTX_EP(comp->op_context);
+	struct mrail_recv *recv;
+
+	// TODO match seq number
+	if (comp->flags & FI_MSG) {
+		FI_DBG(&mrail_prov, FI_LOG_CQ, "Got MSG op\n");
+		fastlock_acquire(&mrail_ep->util_ep.lock);
+		// TODO pass the right address
+		recv = mrail_match_recv_handle_unexp(&mrail_ep->recv_queue, 0,
+						     FI_ADDR_UNSPEC,
+						     (char *)comp, sizeof(*comp),
+						     NULL);
+		fastlock_release(&mrail_ep->util_ep.lock);
+	} else {
+		assert(comp->flags & FI_TAGGED);
+		FI_DBG(&mrail_prov, FI_LOG_CQ, "Got TAGGED op\n");
+		fastlock_acquire(&mrail_ep->util_ep.lock);
+		recv = mrail_match_recv_handle_unexp(&mrail_ep->trecv_queue,
+						     comp->tag,	FI_ADDR_UNSPEC,
+						     (char *)comp, sizeof(*comp),
+						     NULL);
+		fastlock_release(&mrail_ep->util_ep.lock);
```

Comment:
Move the locking outside the if-else statement

Reply:
Done